### ISSUE TITLE
Python environment is missing propagated inputs

### ISSUE THREAD

**Issue by gmacon** [cachix/devenv#601]
**Describe the bug**

The `propagatedBuildInputs` of a Python package are not available in the shell created by devenv.

**To reproduce**

With [this reproducing environment](https://gist.github.com/gmacon/14b1295d54a87418ee24ac33a97a24bd) active, run

```
python -c "import requests"
```

You should get nothing, but instead get:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/nix/store/p1g0gk6dkw7cmgb55mnp56vxksyqhwd2-devenv-profile/lib/python3.10/site-packages/requests/__init__.py", line 43, in <module>
    import urllib3
ModuleNotFoundError: No module named 'urllib3'
```

**Version**
 
devenv: 0.6.2

**Comment by sandydoo** [comment]
I've confirmed the issue. If you're looking for a workaround, you can use `languages.python.package = pkgs.python3.withPackages(ps: [ ps.requests ]);` in the meantime.

**Comment by sandydoo** [comment]
Let's compare 3 different environments:

#### 1. `pkgs.mkShell` without `python.withPackages`
```nix
{
  default = pkgs.mkShell {
    packages = [pkgs.python3.pkgs.requests pkgs.python3];
  };
}
```
`requests` is not added to `site-packages` for this python package. Instead, each python dependency is added to `PYTHONPATH`:

```shell
$ echo $PYTHONPATH
/nix/store/mxgpk88mnqvxll57cyr2x49qlsi0pdc4-python3.10-requests-2.29.0/lib/python3.10/site-packages:/nix/store/n98j6fwy4lml4fc5hhgla3bgbnmrazcl-python3.10-brotlicffi-1.0.9.2/lib/python3.10/site-packages:/nix/store/n2zc7wanmvx51yh7z8cvw0r4bs4fryva-python3.10-cffi-1.15.1/lib/python3.10/site-packages:/nix/store/lmc6b17j60m1p2qqlggq8gw2rd822dxm-python3.10-pycparser-2.21/lib/python3.10/site-packages:/nix/store/92gqjlhr98gf1hv68r6mg44i65a1yfkn-python3-3.10.11/lib/python3.10/site-packages:/nix/store/crlxlwz5ihij0brs21n3r6l1a48y6sqy-python3.10-certifi-2022.12.07/lib/python3.10/site-packages:/nix/store/0p5p1spvq9p2r8lzlziz69b3m3432fra-python3.10-charset-normalizer-3.0.1/lib/python3.10/site-packages:/nix/store/iy9mm79nc0hw7kf6klzah1f4dwxqaflf-python3.10-idna-3.4/lib/python3.10/site-packages:/nix/store/vkpv88slzcgb6rqv7ha57b9yblp44ain-python3.10-urllib3-1.26.14/lib/python3.10/site-packages:/nix/store/vcmc87lniwyzsqbkxi7bb3kv914n8slh-python3.10-brotli-1.0.9/lib/python3.10/site-packages:/nix/store/ahqs5lwjdw7b62y2dxlniamhk0pqswbc-python3.10-pysocks-1.7.1/lib/python3.10/site-packages
```

#### 2. `pkgs.mkShell` with `python.withPackages`

```nix
{
  default = pkgs.mkShell {
    packages =
      let pythonEnv = pkgs.python3.withPackages(ps: [ps.requests]);
      in [ pythonEnv ];
  };
}
```

The python package is wrapped and `site-packages` is populated with the dependencies.
```shell
$ ls /nix/store/n692zy7wl89ni3crl5jsc4hy26n7l0kz-python3-3.10.11-env/lib/python3.10/site-packages/
__pycache__                                     brotli.py                     cffi                                idna-3.4.dist-info        requests                   urllib3
_brotli.cpython-310-aarch64-linux-gnu.so        brotlicffi                    cffi-1.15.1.dist-info               pycparser                 requests-2.29.0.dist-info  urllib3-1.26.14.dist-info
_cffi_backend.cpython-310-aarch64-linux-gnu.so  brotlicffi-1.0.9.2.dist-info  charset_normalizer                  pycparser-2.21.dist-info  sitecustomize.py
_sysconfigdata__linux_aarch64-linux-gnu.py      certifi                       charset_normalizer-3.0.1.dist-info  PySocks-1.7.1.dist-info   socks.py
Brotli-1.0.9.dist-info                          certifi-2022.12.7.dist-info   idna                                README.txt                sockshandler.py
```

#### 3. `devenv.mkShell`

```nix
{
  default = devenv.lib.mkShell {
    inherit inputs pkgs;
    modules = [
      ({ config, ... }: {
        packages = with config.languages.python.package.pkgs; [
          requests
        ];
        languages.python.enable = true;
      })
    ];
  };
}
```

`$PYTHONPATH` points to the python package's `site-packages`. It contains `requests`, but none of the runtime dependencies specified in `propagatedBuildInputs`.

```shell
$ echo $PYTHONPATH
/nix/store/x92y3vc5nfqpnx53paj9bjv1i4mi3mna-devenv-profile/lib/python3.10/site-packages
(devenv)
$ ls /nix/store/x92y3vc5nfqpnx53paj9bjv1i4mi3mna-devenv-profile/lib/python3.10/site-packages
__pycache__  _sysconfigdata__linux_aarch64-linux-gnu.py  README.txt  requests  requests-2.29.0.dist-info  sitecustomize.py

**Comment by sandydoo** [comment]
Modifying https://github.com/cachix/devenv/blob/828b46c011d220172977ccc9d1941d1755a8da98/src/modules/top-level.nix#L9-L13 with the following is enough to fix the issue:

```nix
{
  drvOrPackageToPaths = drvOrPackage:
    (if drvOrPackage ? outputs then
      builtins.map (output: drvOrPackage.${output}) drvOrPackage.outputs
    else
      [ drvOrPackage ]
    ) ++ (drvOrPackage.propagatedBuildInputs or [ ]);
}
```

**Comment by gmacon** [comment]
That's interesting, thanks.

I wondered if we needed to handle the other kinds of propagated inputs, so I took another look at [the Nixpkgs manual section Specifying depedencies](https://nixos.org/manual/nixpkgs/stable/#ssec-stdenv-dependencies). I still don't fully understand, this, but I *think* the answer is that we need `propagatedBuildInputs` , `propagatedNativeBuildInputs`, and `depsTargetTargetPropagated` unconditionally. Does that sound right to folks here?
